File: 프로그래머스_불량 사용자.py
from itertools import combinations, permutations

def is_bad(uid, ban): #아이디가 ban되었는지 체크하는 함수
    if len(uid) != len(ban):
        return False
    else:
        for i,j in zip(uid,ban): #주어진 아이디를 한 글자씩 비교한다.
            if j == '*':
                continue
            if i != j:
                return False
        return True
        
def solution(user_id, banned_id):
    answer = []
    for i in permutations(user_id, len(banned_id)): #ban아이디의 개수만큼 아이디에서 랜덤으로 고름(순열) - 완전탐색
        count = 0
        for a, b in zip(i, banned_id): #주어진 ban아이디 배열과 고른 i순열을 전부 비교함
            if is_bad(a,b): #주어진 아이디가 ban아이디에 부합하면
                count += 1 
        
        if count == len(banned_id) : #부합하는 아이디의 수가 ban아이디의 수와 같고 (= 전부 해당함)
            if set(i) not in answer: #그 아이디 순열이 answer에 포함되지 않은 상태라면
                answer.append(set(i))
                
    return len(answer)

File: 프로그래머스_불량 사용자.cpp
#include <string>
#include <vector>
#include <set>
#include <iostream>
//next_permutation은 너무 느림
using namespace std;

set<string> s; //중복을 제거한 s
bool visited[8] = {false};

bool isBad(string bad, string id) { // 불량 사용자 아이디와 매칭시키기
    if(bad.size() != id.size()) return false;
    
    for(int i=0; i<bad.size(); i++) {
        if(bad[i] == '*') continue;
        if(bad[i] != id[i]) return false;
    }
    return true;
}

void dfs(vector<string>& uid, vector<string>& ban, int idx) {
    if(idx == ban.size()) { //조건에 맞는 아이디를 ban개수만큼 찾았으면
        string tmp;
        for(int i=0; i<uid.size(); i++) {
            if(visited[i]) tmp += uid[i]; //중요한건 s의 사이즈이므로 그냥 합친다. 중복이 있을시 자동으로 삭제된다. 순열이므로 순서가 중요.
        }
        s.insert(tmp);
        return;
    }
    
    for(int i = 0; i< uid.size(); i++) { //uid를 탐색. dfs를 통해 ban.size()개수만큼 뽑아낸다.
        if(visited[i]) continue;     
        if(uid[i].size() != ban[idx].size()) continue;
        
        if(isBad(ban[idx], uid[i])) { //ban의 idx번째 아이디와 비교한다.
            visited[i] = true;
            dfs(uid,ban,idx+1);
            visited[i] = false;
        }
        
    }
}


int solution(vector<string> user_id, vector<string> banned_id) {
    dfs(user_id, banned_id, 0);
    return s.size();
}
